spec/jwt_spec.rb

```diff
@@ -771,4 +771,75 @@
       expect(JWT.decode(token, 'secret', true, algorithm: 'HS256')).to include(payload)
     end
   end
+
+  context 'when multiple algorithms given' do
+    let(:token) { JWT.encode(payload, 'secret', 'HS256') }
+
+    it 'starts trying with the algorithm referred in the header' do
+      expect(::JWT::Algos::Rsa).not_to receive(:verify)
+      JWT.decode(token, 'secret', true, algorithm: ['RS512', 'HS256'])
+    end
+  end
+
+  context 'when algorithm is given as a custom class' do
+    let(:custom_algorithm) do
+      Module.new do
+        module_function
+
+        def sign(*)
+          'custom_signature'
+        end
+
+        def verify(data:, signature:, verification_key:) # rubocop:disable Lint/UnusedMethodArgument
+          signature == 'custom_signature'
+        end
+
+        def alg
+          'custom'
+        end
+
+        def valid_alg?(alg)
+          alg == self.alg
+        end
+      end
+    end
+
+    it 'uses whatever is the implementation' do
+      token = JWT.encode(payload, 'secret', custom_algorithm)
+      expect(token).to eq('eyJhbGciOiJjdXN0b20ifQ.eyJ1c2VyX2lkIjoic29tZUB1c2VyLnRsZCJ9.Y3VzdG9tX3NpZ25hdHVyZQ')
+      expect(JWT.decode(token, 'secret', true, algorithm: custom_algorithm))
```

Comment:
What is actually expected here? Maybe the `payload`?

Reply:
Very good question. A set of new improved tests pushed.